core/src/main/java/com/linecorp/armeria/internal/Http2KeepAliveHandler.java

```diff
@@ -94,22 +92,30 @@
             }
         }
     };
+    private long lastPingPayload;
 
-    Http2KeepAliveHandler(Channel channel, Http2FrameWriter frameWriter, long pingTimeoutInNanos) {
+    public Http2KeepAliveHandler(Channel channel, Http2FrameWriter frameWriter, long pingTimeoutInNanos) {
         this.channel = channel;
         this.frameWriter = frameWriter;
         this.pingTimeoutInNanos = pingTimeoutInNanos;
     }
 
+    private static void throwProtocolErrorException(String msg, Object... args) throws Http2Exception {
+        throw new Http2Exception(Http2Error.PROTOCOL_ERROR, String.format(msg, args));
+    }
+
+    /**
+     * Callback for when the channel is idle.
+     */
     public void onChannelIdle(ChannelHandlerContext ctx, IdleStateEvent event) {
-        checkState(state == State.IDLE, "Waiting for PING ACK or shutdown");
+        checkState(state == State.IDLE, "Invalid state. Expecting IDLE but was " + state);
```

Comment:
`... but was %s", state` so that string concatenation does not occur when state is correct.